spec/derived_build_code_formatter_spec.rb

```diff
@@ -99,22 +211,91 @@
     end
 
     context 'with invalid prefixes' do
-      it 'rejects multi-character strings' do
+      it 'rejects invalid prefix formats' do
+        # Multi-character strings and out-of-range numbers
         expect { described_class.new(prefix: '12') }.to raise_error(/Prefix must be a single digit or empty string/)
-      end
+        expect { described_class.new(prefix: '10') }.to raise_error(/Prefix must be a single digit or empty string/)
+        expect { described_class.new(prefix: '-1') }.to raise_error(/Prefix must be a single digit or empty string/)
 
-      it 'rejects non-numeric strings' do
+        # Non-numeric characters
         expect { described_class.new(prefix: 'a') }.to raise_error(/Prefix must be an integer digit/)
+        expect { described_class.new(prefix: '@') }.to raise_error(/Prefix must be an integer digit/)
       end
+    end
+  end
 
-      it 'rejects numbers outside 0-9 range' do
-        expect { described_class.new(prefix: '10') }.to raise_error(/Prefix must be a single digit or empty string/)
-        expect { described_class.new(prefix: '-1') }.to raise_error(/Prefix must be a single digit or empty string/)
+  describe 'version component validation' do
+    context 'with valid version components within digit limits' do
+      it 'accepts version components that fit within their digit limits' do
+        formatter = described_class.new(major_digits: 1, minor_digits: 2, patch_digits: 2, build_digits: 3)
+
+        # Valid cases: major=9 (1 digit), minor=99 (2 digits), patch=99 (2 digits), build=999 (3 digits)
+        version = Fastlane::Models::AppVersion.new(9, 99, 99, 999)
+        expect { formatter.build_code(version: version) }.not_to raise_error
       end
 
-      it 'rejects symbols and special characters' do
-        expect { described_class.new(prefix: '@') }.to raise_error(/Prefix must be an integer digit/)
-        expect { described_class.new(prefix: '#') }.to raise_error(/Prefix must be an integer digit/)
+      it 'accepts version components at exactly the digit limit' do
+        formatter = described_class.new(major_digits: 2, minor_digits: 1, patch_digits: 3, build_digits: 2)
+
+        # Edge cases: exactly at limits
+        version = Fastlane::Models::AppVersion.new(99, 9, 999, 99)
+        expect { formatter.build_code(version: version) }.not_to raise_error
+      end
+    end
+
+    context 'with invalid version components exceeding digit limits' do
+      it 'rejects major version exceeding digit limit' do
+        formatter = described_class.new(major_digits: 1, minor_digits: 2, patch_digits: 2, build_digits: 2)
+        version = Fastlane::Models::AppVersion.new(10, 1, 1, 1) # major=10 > max(9) for 1 digit
```

Comment:
💄  nitpick: This `> max(9)` in the comment suggests that we're testing the int value, while in practice we're testing the formatted string's length now, so might as well remove that mention? 🤷 

```suggestion
        version = Fastlane::Models::AppVersion.new(10, 1, 1, 1) # major=10 is longer than 1 digit
```
(Of course, same for the other ones below)